Awesome dev - Release of lesson version 1.0 (#1)
carpentries-incubator/scientific-metadata: episodes/structure-schema.Rmd

---

> title: 'Structured metadata: from Markup to JSON'

Reviewer: ```suggestion
title: 'Structured Metadata: From Markup to JSON'
```

<br>

:::::::::::::::::::::::::::::::::::::: questions 

- What is semi-structured metadata?
- How do you extract semi-structured metadata from natural language.
- What is the JSON syntax?

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: objectives

- Explain the importance of semi-structured metadata for machine readability.
- Understand, read and write basic Markdown / HTML / XML / JSON.

::::::::::::::::::::::::::::::::::::::::::::::::

## Markup?

::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: instructor
## Lecture: Introduction to information transfer

**Slide set**: <a href="../files/slideSet/DataTransfertIntro_CarpentriesIncubator_ScientificMetadata.pdf" target="_blank" download>DataTransfertIntro.pdf</a>

**Time estimate**: 5 min

1. With the next episode, we are diving deaper into the topic of structured data and enhancing machine readability of our metadata records.
2. So far we have been talking about natural language communication. When the senior doctor gives a status update about a patient's condition to his colleague, the information is clear. Let's take this message apart and structure it...

> 3. The message is concering a subject, namely the specific patient in room 305. It includes a location, the room number and it gives information about the current condition, unchanged. Let's assume that both doctors were machines. Extracting these information from the natural language text wouldn't be as straight forward...

Reviewer: ```suggestion
3. The message is concering a subject, namely the specific patient in room 305. It includes a location, the room number and it gives information about the current condition, "unchanged". Let's assume that both doctors were machines. Extracting this information from the natural language text wouldn't be as straight-forward...
```

> 5. And the information needs to be structured in particular ways to make it easily transferrable and accessible for the receiving end. One such leightweight data structure is JSON, JavaScript Object Notation. We will take an in depth look at JSON in the upcoming lectures.

Reviewer: ```suggestion
5. And the information needs to be structured in particular ways to make it easily transferrable and accessible for the receiving end. One popular and lightweight format to structure data is JSON, JavaScript Object Notation. We will take an in-depth look at JSON in the upcoming lectures.
```

::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: instructor
## Lecture: What is Markup, XML & JSON

**Slide set**: <a href="../files/slideSet/WhatIsMarkupXmlJson_CarpentriesIncubator_ScientificMetadata.pdf" target="_blank" download>WhatIsMarkupXmlJson.pdf</a>

> **Time estimate**: 15

Reviewer: ```suggestion
**Time estimate**: 15 min
```

> 1. The first term, that we will come across regularly throughout this lesson is Markup.

Reviewer: ```suggestion
1. The first term that we will come across regularly throughout this lesson is *Markup*.
```

> 2. You encounter markup continuously, when you read a text. A period, for example, indicates the end of a declarative or imperative sentence, which results in the speaker briefly stopping to talk, when read out loud. The question mark, on the other hand, declares a question and a speaker will raise their voice before pausing. In this sense, the punctuation - or Markup - is not part of the text or information by itself. It rather give some additional information about the text.

Reviewer: ```suggestion
2. You encounter markup continuously when you read a text. A period, for example, indicates the end of a declarative or imperative sentence, which results in the speaker briefly stopping to talk, if reading out loud. The question mark, on the other hand, declares a question and a speaker will, in many languages, raise their voice before pausing at the end of the sentence. In this sense, the punctuation - or Markup - is not part of the text or information by itself. It rather provides some additional information about the text.
```

> 3. Essential for functional markup is that the sender - in case of a text the writer - and the recipient - the reader - agree on the interpretation of the symbols used to markup the text. Shown here is an interrobang, a combination of questionmark and exclamation point. (<i>You can reactivate the audience by asking them, whether someone is familiar with the definition of the interrobang or if they would know how to interprete it.</i>) The interrobang was proposed in the 1960s as a new punctuation mark to indicate a rhetorical question. As you might have guessed already, the proposal did not come very far.

Reviewer: ```suggestion
3. Essential for functional markup is that the sender, in case of a text - the writer, and the recipient - the reader - agree on the interpretation of the symbols used to markup the text. Shown here is an *interrobang*, a combination of question mark and exclamation point. (<i>You can reactivate the audience by asking them, whether someone is familiar with the definition of the interrobang or if they would know how to interprete it.</i>) The interrobang was proposed in the 1960s as a new punctuation mark to indicate a rhetorical question. As you might have guessed already, the proposal did not come very far.
```

Also: Why HTML `<i>` instead of markdown `*` ? Is it not working?

4. Markup can be distinguished into four different types:
  - punctuational markup, as seen in the previous examples.

>   - presentational markup gives information on how the text should be displayed, which includes vertical or horizontal spacing, page breaks, and enumerations of lists. When enclosing text between two asterisks while writing Markdown, a Markdown interpreter will display this text in bold font face.

Reviewer: "between two asterisks" could be misunderstood as `*this*`, which is *italic* font...

but saying "between two pairs of asterisks" might sound even more confusing... so maybe change the example to the italics one?

EDIT: I forgot there are the slides... if it shows something like `**this**`, ignore this comment.

  - descriptive or declarative markup gives information on the type or class the enclosed text belongs to. In this example, the text was enclosed by an H1 opening and closing tag. HTML interpreters will now know, that this sentence is the most important headline on the page, regardless of how it is displayed.

>   - referential markup refers to entities external to the document. On ta website, this marked up sentence would appear as a link and redirect the user to the specified URL, when activated.

Reviewer: ```suggestion
  - referential markup refers to entities external to the document. On a website, this marked up sentence would appear as a link and redirect the user to the specified URL, when activated.
```

5. I'm certain all of us have manually marked up a document in one way or another, using a text marker while reading a journal publication or commenting a manuscript while reviewing it. These markups give information to the human reader: What is important in this text? Whis information are missing? Markup in computer files, however, is targetting a machine for interpretation. It declares how the content should be formatted or printed, or give additional information on the written words.
6. So from now on in this lecture, the term markup will refer to markup that targets computers, mainly descriptive and referential markup.
7. Descriptive and referential markup, used rigorously, is a way to make data and natural text more accessible for computer analysis and can provide additional information for the human user as well. But we will come to that in a bit.

> 8. We will first shift our focus on the importance of the interoperability of the data formats that we want to use to document our data and metadata. In the 1980s, roughly 10 years before the birth of the world wide web, the internet was already a thing and national and international data networks were emerging fast. This increased the need for standardized data transfer formats to facilitate the data exchange and interpretability on various machines in one networt. One of the first insdustry standards for electronic publishing was the Standard Generalized **Markup** Language or SGML.

Reviewer: ```suggestion
8. We will first shift our focus to the importance of the interoperability of the data formats that we want to use to document our data and metadata. In the 1980s, roughly 10 years before the birth of the World Wide Web, the internet was already a thing and national and international data networks were emerging fast. This increased the need for standardized data transfer formats to facilitate the data exchange and interpretability on various machines in one network. One of the first industry standards for electronic publishing was the "Standard Generalized **Markup** Language", or SGML.
```

10. This XML example stores data in form of natural text (<i>point to the two sentences</i>) and a number (<i>highlight the "1"</i>). Each of these data elements is enclosed in a start and end end tag. As mentioned before, these tags can be defined arbitrarily, so we state, that the first sentence is the title, the second sentence is a description and the number one represents a word count. We can now introduce a data hierarchy, declaring that all these data elements describe one example. We could now look for the "title" of the "example" and retrieve the data stored between those tags. XML is software- and hardware-independent and still widely used as a data exchange format. Though XML is human-readable, enclosing the data between start and end tags is quite an effort and meddles with the clarity of this format. So we are leaving the realm of markup languages...

> 11. to introduce another data exchange format, which is recommended as a standard from the W3 consortium: JSON. JSON is short for JavaScript Object notation. It is inspired by the JavaScript representation for data objects and bears all the advantages XML has shown - like lightweight data exchange, software- and hardware-independency, and hierarchical structure - with an increased human-readability. For the remainder of this lesson, we will learn how to read, write and standardize JSON object literals.

Reviewer: ```suggestion
11. to introduce another data exchange format, which is recommended as a standard from the W3 consortium: JSON. JSON is short for JavaScript Object notation. It is based on the representation for data objects in JavaScript - the main programming language of the Web - and bears all the advantages XML has shown, such as lightweight data exchange, software- and hardware-independence, and hierarchical structure - with an increased human-readability. For the remainder of this lesson, we will learn how to read, write and standardize JSON object literals.
```

Concerning human readability... This might "feel" subjectively true, but very arguable. I'd rather say JSON is easier to **write** by hand - less syntax, less symbols, less concepts. XML with descriptive tags can also be readable. And for any complex object, neither JSON nor XML are fun to work with "by hand" anyway.

So not sure I'd sell the human readability, but rather the *simplicity*. Even techies don't like XML and XSLT, thats why everyone who can jumped ship.

and fyi, something the learners don't need to know, but the second important aspect was probably *pragmatism* - JSON is valid JavaScript **by design**, so JSON is not *inspired* but *based* on JavaScript objects, in a sense being a minimal restricted subset. So **every browser parses JSON for free**, no need to **implement** JSON, it was essentially already **there** , a bit like metadata was formalized as a concept long after it was already used ;) That alone **for sure** helped with the adoption.

::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: instructor

> ## Lecture: JSON step-by-step

Reviewer: ```suggestion
## Lecture: JSON Step-By-Step
```

**Slide set**: <a href="../files/slideSet/JsonStepByStep_CarpentriesIncubator_ScientificMetadata.pdf" target="_blank" download>JsonStepByStep.pdf</a>

**Time estimate**:

1. Let's have a closer look on the structure and syntax of JSON object literals. 
2. In this example we have a JSON object, which describes a person named John.
3. Objects are enclosed in curly braces.
4. Within these objects, data entities are stored in key-value-pairs.

> 5. Keys are always of the data type "string."

Reviewer: ```suggestion
5. Keys are always of the data type "string"
```

6. Data values can be given as one of the following data types. They can be given as...

> 7. strings. 

Reviewer: ```suggestion
7. Strings... 
```

> 8. String values can contain any kind of characters and are characteristically enclosed in parentheses. Each string, no matter how long it is, will be interpreted as a single value.

Reviewer: ```suggestion
8. String values can contain any kind of characters and are typically enclosed in double quotes. Each string, no matter how long it is, will be interpreted as a single value.
```

You mean `"`, not parentheses, right?

9. Numbers...
10. can be either integers or floats.
11. Booleans... 

> 12. Boolean values in JSON has one of two possible values. It is either true or false, or 1 for true and 0 for false, respectively.

Reviewer: ```suggestion
12. Boolean values in JSON have one of two possible values. It is either true or false.
```

I would not tell people to use 0 for false and 1 for true, if we want them to write schemas later... because this kind of inconsistency is actually something we want to avoid. That many places will for convenience sloppily auto-convert things into "truthy" and "falsy" values looks nice, but can be a source of possible errors and make data cleaning hard.

13. Null...

> 14. Assigning the data type NULL does not mean the same as zero but rather that no value was assigned to this variable. You might be familiar with other representations, such as NA or nan. If you read a JSON object into Python, for example, NULL will be parsed as nan.

Reviewer: "will be parsed as nan" - not true. Are you talking about pandas specifically?

`null` naturally maps onto Python `None`, and this is what you get when using most tools....

Author: ```suggestion
14. Assigning the data type NULL does not mean the same as zero but rather that no value was assigned to this variable. You might be familiar with other representations, such as NA or nan. If you read a JSON object into Python, for example, NULL will be parsed as `None`.
```

15. Values in JSON can also be arrays. For example, John has two parents, Anna and Michael.
16. An array is a collection of multiple elements and can be understood as a list of values. The data elements within an array can have the same data type or vary in their data types.
17. Last but not least, a value can be another object. John has a pet and the value of the key "pet" is an object describing John's dog Brutus.

> 18. Again, an object contains key-value pairs. These key-value-paris ar seperated by commata and enclosed by curly braces. 

Reviewer: ```suggestion
18. Again, an object contains key-value pairs. These key-value-pairs are separated by commas and enclosed by curly braces. 
```

> 19. The seperation of data entities by commata in a JSON object literal is crucial.

Reviewer: ```suggestion
19. The separation of data entities by commas in a JSON object literal is crucial.
```

> 20. JSON is not sensitive to indentations and line breaks. However, significant indenting and line breaking, can increase the human readability of the JSON object literal. Let's compare the JSON representation of this object with its XML representation...

Reviewer: ```suggestion
20. JSON is not sensitive to indentations and line breaks. However, indentation and line breaks can significantly increase the human readability of the JSON object literal. Now let's compare the JSON representation of this object with its XML representation...
```

> 21. Both representations contain the same information. In XML the data value is stored between start and end tags, whereas in JSON, the value is assigned to a property and separated from other entries with a comma. Both representations allow for hierarchical data storage as seen for the description of John's pet. In JSON this is done by nesting objects, in XML the data entries are enclosed by `<pet>` tags. A difference can be seen in the representation for John's parents. Whereas an array is given in the JSON object literal, comparable lists do not exist in XML. Here the parents' names are enclosed individually with the same start and end tag.

Reviewer: XML has no arrays, but JSON has no attributes... maybe that should be mentioned

> 22. Besides XML and JSON, the third widely used standard data exchange format is YAML. All three formats can be parsed into one another.

Reviewer: ```suggestion
22. Besides XML and JSON, the third widely used standard data exchange format is YAML. JSON and YAML can be parsed into one another. 
```

Well, yes and no. 

JSON, YAML and TOML really can be "parsed into each other", i.e. are absolutely equivalent ways to write the same objects (ignoring some wild YAML extensions).

XML and the ones above have the mismatch (arrays and attributes) with no standard resolution, which means - ad-hoc mapping decisions. So saying that XML and the others "can be parsed into each other" sounds misleading.

23. (<i>allow for some time for questions</i>)

::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::

**Markup** is not part of the natural text or content but tells something **about it**...[^4], [^5]. 

**Punctuational markup** -- for example -- is placing periods or question marks at the end of sentences.

```text
This is a question?
```

**Presentational markup** is mainly about style. 

```markdown
Simple markdown will interpret the following word **bold**.
```

> 

Reviewer 2: I don't know if the target audience of the course is expected to already be familiar with markdown, but maybe it's valuable to include a sentence or two explaining what it is (a simple standard that can be used to easily markup content)?

Also, I though of a silly markup example that may be familiar to the audience: [In Whatsapp you can format your messages using presentational markup.](https://faq.whatsapp.com/539178204879377?locale=en_US&cms_id=539178204879377&draft=false)

**Descriptive or declarative markup** declares what an element is; e. g. a member of a particular type or class like a:

```markdown
## headline
```

> If design rules for headlines change, the document structure remains intact and is still in line with the authors original intention.

Reviewer: ```suggestion
If design rules for headlines change, the document structure remains intact and is still in line with the authors' original intention.
```

```html
<a href="url">link text displayed to reader</a>
```
:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: callout
Rigorous markup can make text (character strings) more accessible for computer analysis.
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::

**SGML** (**S**tandard **G**eneralized **M**arkup **L**anguage) was one of the first industry standards for electronic publishing -- a meta-language for generalized, descriptive markup languages -- first accepted as an ISO standard in 1986. Both, HTML (1989) and XML (1998) are based on SGML.[^6]

**HTML** (**H**yper**T**ext **M**arkup **L**anguage) is the standard markup language for web pages.

![ [*] **Zender, C.** (2005) "Markup 101: Markup Basics.", SAS Institute. <a href="https://www.lexjansen.com/pharmasug/2005/Tutorials/tu12.pdf" target="_blank">https://www.lexjansen.com/pharmasug/2005/Tutorials/tu12.pdf</a>](fig/cynthiaZenderQuote.png){alt="Quote by Cynthia Zender, SAS Institue, saying 'To make markup work, the writer and the interpreter of the marked up content have to agree on the interpretation of the markup symbols.'"}

## XML

> In contrast, the main purpose of **XML** (e**X**tensible **M**arkup **L**anguage) is **the transfer and storage of arbitrary data** on the World Wide Web.  XML is software- and hardware-independent. It is considered human-readable and allows for hierarchical (tree-like) structures. Data elements are wrapped in start `<...>` and end `</...>` "tags". XML tags can be customized by the author of the document, its markup is extensible.[^7]

Reviewer 2: I would try to emphasize that, even though XML has "markup" in its name, it does not necessarily (if at all) contain markup information about some text.
I can see it being confusing, since the idea of markup was just explained in the previous paragraph.

Reviewer: good point! if not for historical lineage, XML might as well be called "extensible data description language"

Author: I will include this in the talking points for the lecture :)

```xml
<example>
    <title>This is the example title</title>
    <description>A simple XML example</description>
    <wordCount>1</wordCount>
</example>
```

## JSON

**JSON** (**J**ava**S**cript **O**bject **N**otation) is **not** a markup language. It is a **lightweight, human-readable, hierarchical format to store and transport data**.[^8] <a href="https://www.json.org/json-en.html" target="_blank">JSON syntax</a> is inspired by JavaScript object notation.[^9] Like XML, JSON is software- and hardware-independent.

- data elements are defined in key/value pairs
- keys are of data type `string` (in quotes)
- values must be of data type `string`, `number`, `boolean`, `array` or `object` 
- elements are separated by commas
- curly braces hold objects
- square brackets hold arrays
- no comments supported
    
```json
{
    "key":"value",
    "aString":"string",
    "anInteger":5,
    "aFloat":0.5,
    "aBoolean":true,
    "anArray": ["item1", "item2", "item3"],
    "anObject": {
        "key1":"value1",
        "key2":"value2",
        "key3":"value3"
    }
}
```

:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: callout 

> Data exchange formats such as the **markup language XML** or **JSON** can be read and processed not only by humans but also by computers. Structured (meta)data is key to enable **machine-readability**. 

Reviewer: ```suggestion
Data exchange formats such as **XML** or **JSON** can be read and processed not only by humans but also by computers. Structured (meta)data is key to enable **machine-readability**. 
```

If we agree that we should "de-emphasize" the "markup" in the name XML

::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: instructor
## Lecture: Challenge introduction

**Slide set**: <a href="../files/slideSet/Challenge2and3_CarpentriesIncubator_ScientificMetadata.pdf" target="_blank" download>Challenge2and3.pdf</a>

**Time estimate**:

1. In the following challenges, we will apply what we have learned in the lectures so far to our example dataset. 
2.You have inspected the data and realised quite rapidly, that you need more information on the data to make sense of it. So you ask your collaborators for additional meta-information.
3. They reply by sending you a README-style metadata file.
4. However, your funding agency requires that you upload every dataset in your project to their public repository and attach a structured metadata record.
5. In challenge 2, you will review the README-style metadata record and highlight all the information that you deem relevant to understand the data and reproduce the experiment.

> 6. In challenge 3, you will then take these information and transform it into a JSON object literal.

Reviewer: ```suggestion
6. In challenge 3, you will then take this information and transform it into a JSON object literal.
```

7. (<i>The challenges are designed as group challenges. We like to add a slide here, on which we list the randomized groups and give additional technical and administrative information.</i>)

::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: instructor
## Challenge: Identify metadata in README.txt / Write JSON metadata record

The following challenges 2 & 3 will be processed **consecutively in groups of 5 - 6 learners**.

**Total time:** 35 min

**Implementation:** <br>
Prepare shared notes documents for each group of learners attending the lesson. We have good experiences with copying the challenges into the documents and make space for collaborative working on the tasks. <br>

> Try to be present with as many instructors as break-out rooms to moderate the discussions and answer questions.

Reviewer: ```suggestion
Try to be present with as many instructors as there are break-out rooms to moderate the discussions and answer questions.
```

"**t**" describes the **ride time** at which measurements were taken upon activating the recording in **seconds**.
"**ax**" describes the **biomechanical acceleration of the test person on the x axis** in **m/s²**.
"**ay**" describes the **biomechanical acceleration of the test person on the y axis** in **m/s²**.
"**az**" describes the **biomechanical acceleration of the test person on the z axis** in **m/s²**.
"**scr**" is a **boolean** indicating a **detected scream of the test person**.
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::

:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: challenge
## Challenge 3: Write JSON metadata record

You have manually marked up the relevant information in the README. However, your project requires you to provide metadata in the form of a **machine-readable JSON metadata record**. The project provides you with a simple example JSON object (remember, that curly braces hold objects):

```json
{
  "key":"value",

>   "key":"value"

Reviewer: ```suggestion
  "key": "value",
  "key": "value"
```

}
``` 

> - Based on the information identified in the README; write a well-formed, descriptive JSON object.

Reviewer: ```suggestion
- Based on the information identified in the README, write a well-formed, descriptive JSON object.
```

- Collaboratively, find suitable keys (aka variable names) to your values.
- You may want to use some <a href="https://jsonformatter.org/" target="_blank">JSON formatter web service</a> to check and pretty-print your output.

::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: hint

**Keep in mind**, that values in JSON must be one of the following data types:

- a string `""`
- a number `42`
- a boolean `True`
- null `null`
- an array `[]`
- an object `{}`

**Example:**
```json
{
    "key":"value",
    "aString":"string",
    "anInteger":5,
    "aFloat":0.5,
    "aBoolean":true,
    "anArray": ["item1", "item2", "item3"],
    "anObject": {
        "key1":"value1",
        "key2":"value2",

>         "key3":"value3"

Reviewer: ```suggestion
    "key": "value",
    "aString": "string",
    "anInteger": 5,
    "aFloat": 0.5,
    "aBoolean": true,
    "anArray": ["item1", "item2", "item3"],
    "anObject": {
        "key1": "value1",
        "key2": "value2",
        "key3": "value3"
```

    }
}
```
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: solution
This is just one (of many) valid solutions.

```json
{
  "fileName": "trainingObject.csv",
  "abstract": "The data describes the biomechanical acceleration and screams detected of a test person during the ride of the roller coaster \"Flight of the Bat\" in Gotham City.",
  "format": "text/csv",
  "date": "2022-02-28",
  "creator": [
    {
      "creatorName": "Bruce Wayne",
      "creatorAffiliation": "Institute for Vigilance and Nightly Motion - Justice League"
    },
    {
      "creatorName": "Selina Kyle",
      "creatorAffiliation": "Institute for Vigilance and Nightly Motion - Justice League"
    }
  ],
  "experimentalParameters": {
    "testRide": {
      "rideName": "Flight of the Bat",
      "location": "Gotham City, New Jersey",
      "rideType": "roller coaster"
    },
    "testPerson": {
      "sex": "male",

>       "height": 180

Reviewer 2: Even though there's multiple correct solutions, I think that the weight of the test person should be included in the JSON metadata record, since it was marked as valuable information in the previous example.

Also, maybe include units for the height, specially since it was converted from imperial to metric.

      "dataType": "number",
      "columnUnit": "sec"
    },
    {
      "columnName": "ax",
      "columnDescription": "biomechanical acceleration of the test person on the x axis in m/s²",
      "dataType": "number",
      "columnUnit": "m/s²"
    },
    {
      "columnName": "ay",
      "columnDescription": "biomechanical acceleration of the test person on the y axis in m/s²",
      "dataType": "number",
      "columnUnit": "m/s²"
    },
    {
      "columnName": "az",
      "columnDescription": "biomechanical acceleration of the test person on the z axis in m/s²",
      "dataType": "number",
      "columnUnit": "m/s²"
    },
    {
      "columnName": "scr",
      "columnDescription": "boolean indicating a detected scream of the test person",
      "dataType": "boolean",
      "columnUnit": "1"
    }
  ]
}
```
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: instructor
## :coffee: Break :coffee:

We recommend a **15 min break** at this point. <br>

> You can use this time, to set-up the plenary discussion.

Reviewer: ```suggestion
You can use this time to set up the plenary discussion.
```

::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: instructor
## Discussion: Challenge reflection

**Time:** 20 min

**Implementation:** <br> 
Display the final JSON object literals of the different groups next to each other. Allow for some time, so that the learners can compare the different results and identify similarities and differences. 

**Objective:** <br>
At this point, the learners should experience some form of frustration - the metadata is now structured and machine readable, yet interoperability and reusability haven't been achieved. This frustration is relieved in the next episode "**Enabling technolgy & Schemas**", when schemas and standards are introduced.

**Leading questions:**<br>
<ul>

>   <li>What was easy generating the structured metadata record?</li>

Reviewer: ```suggestion
  <li>What was easy while generating the structured metadata record?</li>
```

  <li>What was hard? Which points were intensely discussed in the group?</li>
  <li>Which differences do you see between the different JSON metadata records?</li>
  <li>How do you feel after comparing the results?</li>
</ul>

**Creator's recommendation:** <br>
While moderating the discussion, pay attention to and highlight the following characteristics of the created JSON object literals:
<ul>
  <li>(if done correctly) all JSON object literals are **valid**, **functional** and **machine readable**</li>
  <li>representation of key strings (e.g. camel case vs. snake case, capitalization)</li>
  <li>structure of the JSON object literal (e.g. nested objects vs. flat structure)</li>
  <li>value representation</li>
</ul>

::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: discussion
## Plenary result discussion

> - What was easy generating the structured metadata record?

Reviewer: ```suggestion
- What was easy while generating the structured metadata record?
```

- What was hard? Which points were intensely discussed in the group?
- Which differences do you see between the different JSON metadata records?
- How do you feel after comparing the results?

::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: keypoints 

> - Markups add information to a text that is separated from the content. 

Reviewer: ```suggestion
- Markup languages add information to a text that is separated from the content. 
```

- `XML` and `JSON` are **lightweight**, **hierarchical** file formats to store and transfer data.

> - `XML` and `JSON` are **human readable**, **software** and **hardware independet**

Reviewer: ```suggestion
- `XML` and `JSON` are **human readable**, **software-** and **hardware-independent**
```
